src/sptps_test.c

```diff
@@ -125,6 +134,149 @@ static void usage() {
 	fprintf(stderr, "Report bugs to tinc@tinc-vpn.org.\n");
 }
 
+#ifdef HAVE_MINGW
+
+// Windows does not allow calling select() on anything but sockets. Therefore,
+// to keep the same code as on other operating systems, we have to put a
+// separate thread between the stdin and the sptps loop way below. This thread
+// reads stdin and sends its content to the main thread through a TCP socket,
+// which can be properly select()'ed.
+void *stdin_reader_thread(void *arg) {
+	int *sock_fd = arg;
+	struct sockaddr_in sa;
+	socklen_t sa_size = sizeof(sa);
+
+	while(true) {
+		int peer_fd = accept(*sock_fd, (struct sockaddr *) &sa, &sa_size);
+
+		if(peer_fd < 0) {
+			fprintf(stderr, "accept() failed: %s\n", strerror(errno));
+			continue;
+		}
+
+		if(verbose) {
+			fprintf(stderr, "New connection received from :%d\n", ntohs(sa.sin_port));
+		}
+
+		char buf[1024];
+		ssize_t nread;
+
+		while((nread = read(STDIN_FILENO, buf, sizeof(buf))) > 0) {
+			if(verbose) {
+				fprintf(stderr, "Read %lld bytes from input\n", nread);
+			}
+
+			ssize_t nsend = send(peer_fd, buf, nread, 0);
+
+			if(nsend <= 0) {
+				fprintf(stderr, "Could not send data: %s\n", strerror(errno));
+				break;
+			}
+
+			if(verbose) {
+				fprintf(stderr, "Sent %lld bytes to peer\n", nsend);
+			}
+		}
+
+		shutdown(peer_fd, SD_SEND);
+
+		while(recv(peer_fd, buf, sizeof(buf), 0) > 0) ;
+
+		closesocket(peer_fd);
+	}
+
+	shutdown(*sock_fd, SHUT_RDWR);
+	closesocket(*sock_fd);
+	free(sock_fd);
+}
+
+int start_input_reader() {
+	// Get a random port in range 20000..50000
+	int port = rand() % 30001 + 20000;
```

Comment:
You can bind to port 0, this will let the operating system choose a port. You can then get the address+port you need to `connect()` to it using `getsockname()`.